src/mono/dbi/cordb.cpp

```diff
@@ -615,21 +543,46 @@ MONO_API HRESULT CreateCordbObject(int iDebuggerVersion, void **ppCordb) {
   return S_OK;
 }
 
-HRESULT CordbModule::GetAssembly(
-    /* [out] */ ICorDebugAssembly **ppAssembly) {
-  LOG((LF_CORDB, LL_INFO1000000, "CordbModule - GetAssembly - IMPLEMENTED\n"));
-  *ppAssembly = static_cast<ICorDebugAssembly *>(pAssembly);
-  return S_OK;
+CordbBaseMono::CordbBaseMono(Connection *conn) {
+  this->conn = conn;
+  m_cRef = 0;
+  m_cRefInternal = 0;
+  LOG((LF_CORDB, LL_INFO100000, "%s - New - %p\n", GetClassName(), this));
 }
 
-HRESULT CordbAssembly::GetProcess(
-    /* [out] */ ICorDebugProcess **ppProcess) {
-  LOG((LF_CORDB, LL_INFO1000000,
-       "CorDebugAssembly - GetProcess - IMPLEMENTED\n"));
-  *ppProcess = static_cast<ICorDebugProcess *>(pProcess);
-  return S_OK;
+CordbBaseMono::~CordbBaseMono() {}
+
+ULONG CordbBaseMono::InternalAddRef() {
+  return InterlockedIncrement(&m_cRefInternal);
+}
+
+ULONG CordbBaseMono::InternalRelease() {
+  ULONG cRef = InterlockedDecrement(&m_cRefInternal);
+  LOG((
+      LF_CORDB, LL_INFO100000,
+      "%s - InternalRelease - Decrement - EXTERNAL - %d - INTERNAL - %d - %p\n",
+      GetClassName(), m_cRef, cRef, this));
+  if (cRef == 0 && m_cRef == 0) {
+    LOG((LF_CORDB, LL_INFO100000, "%s - InternalRelease - Delete - %p\n",
+         GetClassName(), this));
+    delete this;
+  }
+  return cRef;
 }
 
-CordbBaseMono::CordbBaseMono(Connection *conn) { this->conn = conn; }
+ULONG CordbBaseMono::BaseAddRef() { return InterlockedIncrement(&m_cRef); }
+
+ULONG CordbBaseMono::BaseRelease() {
+  ULONG cRef = InterlockedDecrement(&m_cRef);
+  LOG((LF_CORDB, LL_INFO100000,
+       "%s - Release - Decrement - EXTERNAL - %d - INTERNAL - %d - %p\n",
+       GetClassName(), cRef, m_cRefInternal, this));
+  if (cRef == 0 && m_cRefInternal == 0) {
```

Comment:
This (and `CordbBaseMono::InternalRelease` doing the same thing) is  pretty dangerous and could be a race condition.  One thread could be decrementing the refcount while another is incrementing the internal refcount, for example.

In general working with 2 atomic counters is tricky.

One thing we do elsewhere in Mono is to stuff 2 32-bit counters into one 64-bit value:
```c
  int64_t two_counters;
  int32_t counter_one (int64_t two_counters) { return (int32_t)(two_counters >> 32); }
  int32_t counter_two (int64_t two_counters) { return (int32_t)(two_counters & 0xFFFFFFFF); }
```

Then to atomically decrement use a compare and exchange in a loop:
```c
int64_t old_val;
int64_t new_val;
do {
   old_val = two_counters;
   new_val = ((int64_t)counter_one (old_val)) << 32 | (counter_two (old_val) - 1);
while (CompareExchange64 (&two_counters, new_val, old_val) != old_val);
bool both_zero =  (new_val == 0);
if (both_zero) {
   /* dealloc */
}    
```

Reply:
I can use only one, I used two because I was investigating leaks. I will fix it.

Reply:
Does this need to be lockfree?

Does it need two refcounts?

Reply:
No, we don't need two... I did it only to help me to find if I was missing references or vsdbg was missing references.

Reply:
Do we know that there are no cycles with internal references?

Reply:
Yes, I know that there are no cycles. 
I tested a lot, also used external tools to check that there is no leak. Everything is being deleted in the end.
Do you think there is any other way to check?

Reply:
I believe, with the exception of a handful of API (maybe just 1 or 2), the APIs are meant to be called on the same thread. @tommcdon to confirm.